File: javascript/packages/linter/src/rules/a11y-no-redundant-image-alt.ts
import { BaseRuleVisitor } from "./rule-utils.js"
import { getAttribute, getAttributeValue, hasAttributeValue, getTagLocalName, isHTMLOpenTagNode, isERBOpenTagNode, filterHTMLAttributeNodes, findAttributeByName, hasDynamicAttributeValue } from "@herb-tools/core"

import { ParserRule } from "../types.js"
import type { UnboundLintOffense, LintContext, FullRuleConfig } from "../types.js"
import type { HTMLElementNode, HTMLAttributeNode, ParseResult, ParserOptions } from "@herb-tools/core"

const REDUNDANT_ALT_WORDS = new Set(["image", "picture"])

class NoRedundantImageAltVisitor extends BaseRuleVisitor {
  visitHTMLElementNode(node: HTMLElementNode): void {
    this.checkImgTag(node)
    super.visitHTMLElementNode(node)
  }

  private checkImgTag(node: HTMLElementNode): void {
    const tagName = getTagLocalName(node)

    if (tagName !== "img") {
      return
    }

    const altAttribute = this.getAltAttribute(node)

    if (!altAttribute) return
    if (!hasAttributeValue(altAttribute)) return
    if (hasDynamicAttributeValue(altAttribute)) return

    const altValue = getAttributeValue(altAttribute)

    if (!altValue) return

    const words = altValue.toLowerCase().split(/\s+/)

    if (words.some(word => REDUNDANT_ALT_WORDS.has(word))) {
      this.addOffense(
        "`<img>` `alt` prop should not contain \"image\" or \"picture\" as screen readers already announce the element as an image.",
        altAttribute.location,
      )
    }
  }

  private getAltAttribute(node: HTMLElementNode): HTMLAttributeNode | null {
    const openTag = node.open_tag

    if (isHTMLOpenTagNode(openTag)) {
      return getAttribute(openTag, "alt")
    }

    if (isERBOpenTagNode(openTag)) {
      return findAttributeByName(filterHTMLAttributeNodes(openTag.children), "alt")
    }

    return null
  }
}

export class A11yNoRedundantImageAltRule extends ParserRule {
  static ruleName = "a11y-no-redundant-image-alt"
  static introducedIn = this.version("0.9.7")

  get defaultConfig(): FullRuleConfig {
    return {
      enabled: false,
      severity: "warning"
    }
  }

  get parserOptions(): Partial<ParserOptions> {
    return { action_view_helpers: true }
  }

  check(result: ParseResult, context?: Partial<LintContext>): UnboundLintOffense[] {
    const visitor = new NoRedundantImageAltVisitor(this.ruleName, context)
    visitor.visit(result.value)
    return visitor.offenses
  }
}

File: rust/herb-linter/src/rules/a11y_no_redundant_image_alt.rs
use crate::utils::tag_utils::{get_attribute_value, get_element_attribute, get_tag_local_name, has_attribute_value, has_dynamic_attribute_value};

use herb::nodes::HTMLElementNode;
use herb::Visitor;

const REDUNDANT_ALT_WORDS: &[&str] = &["image", "picture"];

rule_visitor!(NoRedundantImageAltVisitor);
define_parser_rule!(
  A11yNoRedundantImageAltRule,
  "a11y-no-redundant-image-alt",
  Warning,
  NoRedundantImageAltVisitor,
  enabled: false,
  parser_options: { action_view_helpers: true },
  introduced_in: "0.10.2"
);

impl Visitor for NoRedundantImageAltVisitor {
  fn visit_html_element_node(&mut self, node: &HTMLElementNode) {
    if get_tag_local_name(node).as_deref() == Some("img") {
      if let Some(alt_attribute) = get_element_attribute(node, "alt") {
        if has_attribute_value(alt_attribute) && !has_dynamic_attribute_value(alt_attribute) {
          if let Some(alt_value) = get_attribute_value(alt_attribute) {
            let is_redundant = alt_value.to_lowercase().split_whitespace().any(|word| REDUNDANT_ALT_WORDS.contains(&word));

            if !alt_value.is_empty() && is_redundant {
              self.add_offense(
                "`<img>` `alt` prop should not contain \"image\" or \"picture\" as screen readers already announce the element as an image.",
                alt_attribute.location.clone(),
              );
            }
          }
        }
      }
    }

    self.walk_html_element_node(node);
  }
}
